### PR TITLE
BUG: Fix localize_pydatetime using meta datetimes as Timestamp (#25734)

### LINKED ISSUE #25734: Pandas 0.24.x throws an error when using freezegun with CustomBusinessDay
I originally opened an issue with Freezegun:
https://github.com/spulec/freezegun/issues/284

**However, this appears to be a new quirk of Pandas 0.24's `CustomBusinessDay` class and its use of `pandas._libs.tslibs.conversion.localize_pydatetime()`.** Other offset functions don't seem to be affected, for the ones I've tried.

Everything worked just fine with Pandas 0.23.4 and earlier, with the latest freezegun-0.3.11.

#### Code Sample, a copy-pastable example if possible

Test setup:  Python 2.7 and 3.6 on OS X using freezegun-0.3.11 with pandas-0.24.2

```python

import pytz
import pandas
from datetime import datetime
from pandas.tseries.offsets import (CustomBusinessDay, BusinessDay)
from freezegun import freeze_time

LOCAL_TZ = pytz.timezone('America/Los_Angeles')

@freeze_time(LOCAL_TZ.localize(datetime(2019, 3, 15, 0, 0, 0)))
def these_work():
    pandas.Timestamp(datetime.now(LOCAL_TZ))
    pandas.Timestamp(datetime.now(LOCAL_TZ)) + BusinessDay(n=1)

@freeze_time(LOCAL_TZ.localize(datetime(2019, 3, 15, 0, 0, 0)))
def this_fails():
    pandas.Timestamp(datetime.now(LOCAL_TZ)) + CustomBusinessDay(n=1)

these_work()
this_fails()

```
#### Problem description

Failure mode:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-452eb6a5f720> in <module>
     17
     18 these_work()
---> 19 this_fails()

~/.virtualenvs/python3/lib/python3.6/site-packages/freezegun/api.py in wrapper(*args, **kwargs)
    656                     result = func(time_factory, *args, **kwargs)
    657                 else:
--> 658                     result = func(*args, **kwargs)
    659             return result
    660         functools.update_wrapper(wrapper, func)

<ipython-input-2-452eb6a5f720> in this_fails()
     14 @freeze_time(LOCAL_TZ.localize(datetime(2019, 3, 15, 0, 0, 0)))
     15 def this_fails():
---> 16     pandas.Timestamp(datetime.now(LOCAL_TZ)) + CustomBusinessDay(n=1)
     17
     18 these_work()

pandas/_libs/tslibs/offsets.pyx in pandas._libs.tslibs.offsets.BaseOffset.__radd__()

pandas/_libs/tslibs/offsets.pyx in pandas._libs.tslibs.offsets._BaseOffset.__add__()

~/.virtualenvs/python3/lib/python3.6/site-packages/pandas/tseries/offsets.py in wrapper(self, other)
     70
     71             if self._adjust_dst:
---> 72                 result = conversion.localize_pydatetime(result, tz)
     73
     74             result = Timestamp(result)

pandas/_libs/tslibs/conversion.pyx in pandas._libs.tslibs.conversion.localize_pydatetime()

pandas/_libs/tslibs/conversion.pyx in pandas._libs.tslibs.conversion.localize_pydatetime()

AttributeError: 'FakeDatetime' object has no attribute 'tz_localize'
```

#### Expected Output

All of these should work equally well, including `CustomBusinessDay`.

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.8.final.0
python-bits: 64
OS: Darwin
OS-release: 18.2.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: None
pip: 19.0.3
setuptools: 40.6.3
Cython: 0.29.3
numpy: 1.16.0
scipy: 1.2.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.3
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: 4.3.2
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
```

</details>

### PR DESCRIPTION
**Pull request by ArtificialQualia**
- [x] closes #25734
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Proposed fix for #25734

The issue was due to localize_pydatetime assuming anything that was `not PyDateTime_CheckExact` must be a pandas `Timestamp`.  However, in the rare case a `datetime` was created with a custom metaclass, that `datetime` would fail `PyDateTime_CheckExact`, but obviously was not a `Timestamp`.

Therefore, I made the `not PyDateTime_CheckExact` check more explicitly look for a `Timestamp`.  As the `Timestamp` class could not be imported for a `isinstance` check due to a circular import, I took the route that many of the other `is_*` style checks took of adding a `_typ` property to `Timestamp` and inspecting that to determine identity.

### CODE REVIEW

File: pandas/_libs/tslibs/conversion.pyx
Comment (mroeschke) [inline pandas/_libs/tslibs/conversion.pyx]: Instead of this could you `try/except` the `dt.tz_localize(tz)` call, and on an `AttributeError` let branch below handle this?

File: pandas/tests/tslibs/test_conversion.py
Comment (mroeschke) [inline pandas/tests/tslibs/test_conversion.py]: Could you test the `CustomBusinessDay` day arithmetic in the original issue as well?

File: doc/source/whatsnew/v0.25.0.rst
Comment (mroeschke) [inline doc/source/whatsnew/v0.25.0.rst]: This should describe the main user facing impact that was in the original issue; arithmetic with `CustomBusinessDay` and datetime subclasses.

**Review by mroeschke — CHANGES_REQUESTED** [review]
Some comments on the fix

**Comment by mroeschke** [comment]
This looks to be a regression in 0.24 (from this PR https://github.com/pandas-dev/pandas/pull/21691) and eligible for backporting.

File: pandas/_libs/tslibs/conversion.pyx
Comment (ArtificialQualia) [reply pandas/_libs/tslibs/conversion.pyx]: Yeah, that does fix the issue as well.  Do you think that is better than checking for the correct type?

Comment (mroeschke) [reply pandas/_libs/tslibs/conversion.pyx]: I think it's better since it adds less code and `PyDateTime_CheckExact` is supposed to be a quicker check for `Timestamp` objects which we expect to see here more often.

**Comment by mroeschke** [comment]
Will need to `isort` this file.

`ERROR: /home/vsts/work/1/s/pandas/tests/tslibs/test_conversion.py Imports are incorrectly sorted.`

File: pandas/tests/tslibs/test_conversion.py
Comment (ArtificialQualia) [reply pandas/tests/tslibs/test_conversion.py]: Good call on this, in testing the actual arithmetic with a hand-made custom datetime I found a similar problem in `conversion.pyx:convert_datetime_to_tsobject` as well.  It wasn't being hit when testing with the library the user was using as their custom datetimes just happen to implement the `nanosecond` property that regular `datetime` doesn't have.

I'll include that fix in this PR as well if there is no issue with that, since it is the same issue and it was just a quirk of `freezegun` that it didn't show up in that issue.

**Review by jreback — CHANGES_REQUESTED** [review]
pls show if this causes any perf issues

File: pandas/tests/tslibs/test_conversion.py
Comment (jbrockmendel) [inline pandas/tests/tslibs/test_conversion.py]: Why not FakeDatetime(datetime)?

Comment (ArtificialQualia) [reply pandas/tests/tslibs/test_conversion.py]: I was trying to more accurately match what was in the original issue, which used a metaclass, but this issue does show up in the same way on both regular inheritance and classes that use a metaclass.

I can change this to a regular class if desired, it would be testing the same behavior.

File: pandas/tests/arithmetic/test_datetime64.py
Comment (ArtificialQualia) [inline pandas/tests/arithmetic/test_datetime64.py]: A metaclass is needed in this test case to be able to override __instancecheck__

File: pandas/tests/arithmetic/test_datetime64.py
Comment (ArtificialQualia) [inline pandas/tests/arithmetic/test_datetime64.py]: At least some level of monkeypatching is necessary for the original issue to show up.

However, this entire `for` block could be replaced with `m.setattr("pandas.tseries.offsets.datetime", FakeDatetime)`, but I wanted to future proof this test better and replicate what `freezegun` and other libraries that might monkeypatch `datetime` would do.

File: doc/source/whatsnew/v0.25.0.rst
Comment (mroeschke) [inline doc/source/whatsnew/v0.25.0.rst]: `arithmetic`

File: pandas/_libs/tslibs/conversion.pyx
Comment (mroeschke) [inline pandas/_libs/tslibs/conversion.pyx]: Can you comment that this branch is for datetime subclasses?

File: pandas/_libs/tslibs/conversion.pyx
Comment (mroeschke) [inline pandas/_libs/tslibs/conversion.pyx]: Same

**Comment by ArtificialQualia** [comment]
> pls show if this causes any perf issues

I ran the perf testing and did have some significant deviations with many of them in places I wouldn't expect to see any differences.  I'll run them again tonight and see if the deviations are consistent.

File: pandas/tests/arithmetic/test_datetime64.py
Comment (mroeschke) [inline pandas/tests/arithmetic/test_datetime64.py]: Could you expand this comment to describe what `freezegun` (and other libraries) are doing?

**Review by mroeschke — CHANGES_REQUESTED** [review]
One final comment clarification note from me otherwise LGTM.

Okay to backport to the 0.24 branch @jreback? This was a regression in 0.24.

**Comment by jreback** [comment]
no on the backport and this need perf checking

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: we don't throw an exception, rather we raise an exception. Can you even be more explicit here, what does this have to do with CBD?

**Review by jreback — CHANGES_REQUESTED** [review]
I am not clear what problem you are trying to solve here. can you elaborate.

**Comment by jbrockmendel** [comment]
> I ran the perf testing and did have some significant deviations with many of them in places I wouldn't expect to see any differences. I'll run them again tonight and see if the deviations are consistent.

Yah this is a common issue with `asv` when looking at cython code.  If it persists, just show some relevant `%timeit` results.

Because this is replacing a C function call with a python attribute lookup, I'd expect the performance to take a hit.  The question is whether this is called enough times for that perf hit to be non-trivial.

Also, there are a handful of places where we use `not Datetime_CheckExact(x)` as a proxy for `isinstance(x, Timestamp)`.  Ideally if we fix one of these, we should fix them all.

**Comment by ArtificialQualia** [comment]
**%timeit Results**

For the performance tests that over both runs had a factor > 1.1 (did not test the factor < 0.9) here are the %timeit results:

**`multiindex_object.GetLoc.time_large_get_loc`**
This PR:
```
In [4]: %timeit get_loc.time_large_get_loc()
11.8 µs ± 52.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
Master:
```
In [4]: %timeit get_loc.time_large_get_loc()
11.7 µs ± 30 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

**`inference.NumericInferOps.time_subtract(<class 'numpy.int16'>)`**
This PR:
```
In [5]: %timeit nio.time_subtract(np.int16)
431 µs ± 2.31 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
Master:
```
In [5]: %timeit nio.time_subtract(np.int16)
428 µs ± 1.94 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**`timeseries.ResampleSeries.time_resample('datetime', '5min', 'mean')`**
This PR:
```
In [4]: %timeit rs.time_resample('datetime', '5min', 'mean')
3.68 ms ± 56.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
Master:
```
In [4]: %timeit rs.time_resample('datetime', '5min', 'mean')
3.78 ms ± 41.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**`series_methods.NSort.time_nsmallest('last')`**
This PR:
```
In [4]: %timeit ns.time_nsmallest('last')
2.04 ms ± 7.51 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
Master:
```
In [4]: %timeit ns.time_nsmallest('last')
2.3 ms ± 13.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**`multiindex_object.Integer.time_get_indexer`**
This PR:
```
In [4]: %timeit i.time_get_indexer()
1.27 ms ± 9.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
Master:
```
In [4]: %timeit i.time_get_indexer()
1.28 ms ± 2.73 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Out of these, the only 'significant' result was `series_methods.NSort.time_nsmallest('last')` which is a little faster in this PR's code, but that's probably a fluke.

File: doc/source/whatsnew/v0.25.0.rst
Comment (ArtificialQualia) [reply doc/source/whatsnew/v0.25.0.rst]: I tried to scope this description to more what was in the original issue, but really it is due to `freezegun` monkeypatching `datetime` with a subclass of `datetime`.  A few of the `offsets.py` classes turn things into `datetime` in their `apply` functions.  That's why this affects `CustomBusinessDay` and not, for instance, `BusinessDay`.  After a brief look, a few other classes might be affected by this bug, like `Easter` and `FY5253`.

However, in writing a custom `datetime` class for testing, another bug was discovered in `convert_datetime_to_tsobject` which prevents *any* subclassed `datetime` from being converted into a `Timestamp`.  `freezegun` was not affected by this by chance because it happened to implement the `nanosecond` property in their custom `datetime` class, but that fix has been included in this PR as mentioned in another comment.

**Comment by jbrockmendel** [comment]
For the %timeit checks, I was suggesting that you run just the affected code, e.g `%timeit localize_pydatetime(whatever)` (for values that would be affected by the PR).  Or if there are other functions that call localize_pydatetime a lot in a loop, time those.

**Comment by ArtificialQualia** [comment]
> Also, there are a handful of places where we use `not Datetime_CheckExact(x)` as a proxy for `isinstance(x, Timestamp)`. Ideally if we fix one of these, we should fix them all.

I'm happy to do that, but should that be done in a separate issue/PR?

Also note that another option for all those cases would be to replace `PyDateTime_CheckExact` with `PyDateTime_Check`, but that may be less efficient than the `try, except` blocks.

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: ok, so say this affects subclasses of datetime. note btw, we don't provide *any* compat with this, but as usual will take patches.

**Review by jreback — CHANGES_REQUESTED** [review]
so rather than using PyDateTime_CheckExact, if we switch this to PyDateTime_Check it will likely work, though IIRC we did this as a perf issue? @jbrockmendel do you remember

**Comment by ArtificialQualia** [comment]
> so rather than using PyDateTime_CheckExact, if we switch this to PyDateTime_Check it will likely work, though IIRC we did this as a perf issue? @jbrockmendel do you remember

I thought that might work, but since we are using `not PyDateTime_CheckExact` as a proxy for `isinstance(x, Timestamp)`, it wouldn't work to switch that out for `PyDateTime_Check` since Timestamp is itself a subclass of datetime.

File: doc/source/whatsnew/v0.25.0.rst
Comment (ArtificialQualia) [reply doc/source/whatsnew/v0.25.0.rst]: The fact that this affects datetime subclasses is already in the release notes, I'm not sure what you would like changed.

**Comment by jbrockmendel** [comment]
> so rather than using PyDateTime_CheckExact, if we switch this to PyDateTime_Check it will likely work, though IIRC we did this as a perf issue?

@ArtificialQualia's comment about `not PyDateTime_CheckExact(x)` being a proxy for `isinstance(x, Timestamp)` is correct, as is @jreback's point that we are using this proxy for performance reasons.

An ideal solution would be to implement a `is_timestamp` function with C performance.  Some speculation about what that might look like:
- if we somehow merged `Timestamp` with `_Timestamp`, so that `Timestamp` itself were a cdef class, then I _think_ `type(x) is Timestamp` would be a pointer comparison, i.e. fast.  Might merit checking on the cython mailing list about this.
- this might require a refactor to avoid circular dependencies.  Presumably such an `is_timestamp` would be defined in `tslibs.timestamps` (which depends on `tslibs.conversion`), but ATM we use this heuristic in `tslibs.conversion` (also`tslibs.timedeltas` and `tslib`)

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: > :class:`CustomBusinessDay` arithmetic is misleading

File: pandas/_libs/tslibs/conversion.pyx
Comment (jreback) [inline pandas/_libs/tslibs/conversion.pyx]: why is this probably?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
pinged the cython mailing list about fast isinstance checks:  https://groups.google.com/forum/#!topic/cython-users/AQSGs-Sx9Pc

It sounds like `isinstance(x, _Timestamp)` might be significantly faster than `isinstance(x, Timestamp)`

**Comment by jbrockmendel** [comment]
After doing some profiling, it looks like the try/except AttributeError check is _way_ slower than basically every other alternative:

Within cython, had a function run each of the following 10k times.  Then ran that function in `%timeit`

```
                                 Timestamp     datetime
isinstance(x, _Timestamp) -->    16.1µs        17.3µs
isinstance(x, Timestamp)  -->    74.6µs        198µs
not PyDateTime_CheckExact(x) --> 119ns         113 ns
try/except  -->                  3.22 ms       3.21 ms

```

So the best "do it right" (`isinstance(x, _Timestamp)`) is 135x slower than the current check, but 200x better than the try/except check.

There are a couple of extant places where we use this check.  I'm going to make a PR to get rid of those.  @ArtificialQualia are you OK with putting this on hold until a performant solution is in place?

**Comment by ArtificialQualia** [comment]
@jbrockmendel I tried to replicate your results based on your testing criteria, and got some similar results with some exceptions.  For instance:
- `isinstance(x, _Timestamp)` was as fast as `not PyDateTime_CheckExact(x)` within a few nanoseconds.
- `try/except` was only in the `ms` order of magnitude when it actually *catches* an exception.  When no exception was raised (which would almost always be the case in this code), it was only about 2x slower than `isinstance(x, _Timestamp)`

Either way, it seems to me as well the answer is to use `isinstance(x, _Timestamp)`.

I'm happy to put this on hold until your other PR is merged.

**Comment by jbrockmendel** [comment]
> I tried to replicate your results based on your testing criteria, and got some similar results with some exceptions

When I tried this, I found the signature of the `is_timestamp` functions affected the results quite a bit.  In particular, the timings I reported all had the signature `cdef bint is_timestamp(datetime obj)` as opposed to `cdef bint is_timestamp(object obj)`.  Can that account for the difference in results?

> I'm happy to put this on hold until your other PR is merged.

My first attempt at this failed and I don't know when I'll be able to circle back to it.  Better not to wait on me.

**Comment by ArtificialQualia** [comment]
> When I tried this, I found the signature of the `is_timestamp` functions affected the results quite a bit. In particular, the timings I reported all had the signature `cdef bint is_timestamp(datetime obj)` as opposed to `cdef bint is_timestamp(object obj)`. Can that account for the difference in results?

My signature was `cpdef inline void test_1(datetime dt)`, so there are some differences but the arg was the same at least.

I found a very large difference depending on how `_Timestamp` was imported.  If it was imported with a python `import` then 10,000 `isinstance(dt, _Timestamp)` took `55.7 µs` and `167 µs` for Timestamp and datetime respectively.  But if I did a `cimport _Timestamp` (requires additional .pxd changes of course) then that went back down to `70.2 ns` and `64.5 ns` respectively.  Maybe that is where the difference is?  Which import method did you use?

I'm happy to make the PR for the larger scope of changes, but if you were thinking of having someone else do it that is fine with me as well.

**Comment by ArtificialQualia** [comment]
#25853 implements this

**Pull request closed without merging**
